src/lib/draw-and-pick.js

```diff
@@ -53,91 +56,88 @@ export function pickLayers(gl, {
     scissorTest: {x: deviceX, y: deviceY, w: 1, h: 1}
   }, () => {
 
-    let layerIndex = 0;
-    let zOrder = 0;
-
-    for (let i = layers.length - 1; i >= 0; --i) {
-      const layer = layers[i];
-
-      if (layer.props.visible) {
-        layerIndex++;
-      }
-
+    // Clear the frame buffer
+    gl.clear(GL.COLOR_BUFFER_BIT | GL.DEPTH_BUFFER_BIT);
+    // Save current blend settings
+    const oldBlendMode = getBlendMode(gl);
+    // Set blend mode for picking
+    // always overwrite existing pixel with [r,g,b,layerIndex]
+    gl.enable(gl.BLEND);
+    gl.blendFuncSeparate(gl.ONE, gl.ZERO, gl.CONSTANT_ALPHA, gl.ZERO);
+    gl.blendEquation(gl.FUNC_ADD);
+
+    // Render all pickable layers
+    layers.forEach((layer, layerIndex) => {
       if (layer.props.visible && layer.props.pickable) {
 
-        // Clear the frame buffer, render and sample
-        gl.clear(GL.COLOR_BUFFER_BIT | GL.DEPTH_BUFFER_BIT);
-        const info = createInfo({
-          layer,
-          pixel: [x, y],
-          devicePixel: [deviceX, deviceY],
-          pixelRatio
-        });
+        // Encode layerIndex with alpha
+        gl.blendColor(0, 0, 0, (layerIndex + 1) / 255);
 
-        layer.pickLayer({
-          info,
-          uniforms: Object.assign({},
+        layer.drawLayer({
+          uniforms: Object.assign(
+            {renderPickingBuffer: 1, pickingEnabled: 1},
             layer.context.uniforms,
             getUniformsFromViewport(layer.context.viewport, layer.props),
             {layerIndex}
-          ),
-          pickEnableUniforms: {renderPickingBuffer: 1, pickingEnabled: 1},
-          pickDisableUniforms: {renderPickingBuffer: 0, pickingEnabled: 0},
-          deviceX, deviceY,
-          mode
+          )
         });
+      }
+    });
+
+    // Read color in the central pixel, to be mapped with picking colors
+    const color = new Uint8Array(4);
+    gl.readPixels(deviceX, deviceY, 1, 1, GL.RGBA, GL.UNSIGNED_BYTE, color);
+
+    // Decode alpha to layer index
+    const pickedLayer = layers[color[3] - 1];
+    const baseInfo = Object.assign(
+      createInfo([x, y], viewport),
+      {
+        devicePixel: [deviceX, deviceY],
+        pixelRatio
+      }
+    );
+
+    layers.forEach(layer => {
```

Comment:
>  Since we already know which layer is picked, why we still want to loop over all layers and send the picking info every unpicked layer to decode?

One tricky part is letting other layers know that they are no longer picked (so that they can remove highlights etc). We either have to remember which layer was picked last time and send a deselect to that layer, or just keep informing all layers that they are not picked. Some comments to this effect would be good.